botocore/loaders.py

```diff
@@ -221,6 +242,90 @@ def create_loader(search_path_string=None):
     return Loader(extra_search_paths=paths)
 
 
+def _create_path(path):
+    """Construct a BotoZipPath or pathlib.Path object from a path string.
+
+    Split the root (path to a zip) and the path to the subdirectory/file e.g.
+    '/Users/user123/Documents/Archive.zip/data' -->
+    BotoZipPath('/Users/user123/Documents/Archive.zip', '/data')
+    If the zipped path doesn't exist, return a pathlib.Path object instead
+
+    :type path: str
+    :param path: Full path to construct a BotoZipPath or pathlib.Path object
+
+    :return: A BotoZipPath/pathlib.Path object consisting of the root and
+    the path to a file or subdirectory
+
+    """
+    parts = path.split(os.sep)
+    new_path = os.sep
+    while len(parts) > 0:
+        new_path = os.path.join(new_path, parts.pop(0))
+        if is_zipfile(new_path):
+            return BotoZipPath(new_path).joinpath(*parts)
+    return pathlib.Path(path)
+
+
+class BotoZipPath(ZipPath):
+    """A wrapper for ZipPath. This allows for zipped botocore to work with
+    python 3.9. See https://github.com/python/cpython/issues/86256 for
+    additional details.
+
+    """
+
+    def joinpath(self, *other):
+        next = posixpath.join(self.at, *other)
+        return self._next(self.root.resolve_dir(next))
+
+    def _next(self, at):
+        return self.__class__(self.root, at)
```

Comment:
added this because `joinpath` was returning a `zipfile.Path` type instead of `BotoZipPath`, which caused some tests to fail.